File: Data Structures & Algorithms/jump-game/submission-3.py
class Solution:
    def canJump(self, nums: List[int]) -> bool:
        """
            1. lets do brute force first think through the solution
            2. now lets cache this
            3. now lets do bottom up O(n^2)


        """
        # cache = {}
        # def helper(idx,nums: List[int]) -> bool:
        #     if idx >= len(nums)-1:
        #         return True
            
        #     if idx in cache:
        #         return cache[idx]
            
        #     ans = False
        #     for i in range(1, nums[idx]+1):
        #         ans |= helper(idx + i, nums)

        #     cache[idx] = ans
        #     return ans
        

        # return helper(0,nums)
        n = len(nums)
        dp = [False] * n
        dp[0] = True

        for i in range(1,n):
            for j in range(i):
                if j + nums[j] >= i:
                    dp[i] = True
                    break
        
        ans = True
        for i in dp:
            ans &= i
        
        return ans

File: Data Structures & Algorithms/interleaving-string/submission-3.cpp
class Solution {
public:
    vector<vector<int>>cache;
    bool helper(string s1,string s2,string s3,int i,int j){
        if(i + j >= s3.size()){
            return (i >= s1.size() && j>= s2.size());
        }

        if(cache[i][j] != -1){
            return cache[i][j];
        }
        
        bool ans = false;
        if(s1[i] == s3[i+j]){
            ans |= helper(s1,s2,s3,i+1,j);
        }
        if(s2[j] == s3[i+j]){
            ans |= helper(s1,s2,s3,i,j+1);
        }
        return cache[i][j] = ans;
    }

    bool isInterleave(string s1, string s2, string s3) {
        int n = s1.size();
        int m = s2.size();
        cache.resize(n+1,vector<int>(m+1,-1));
        return helper(s1,s2,s3,0,0);
    }
};
